util/coverage/coverage_off_target.py

```diff
@@ -0,0 +1,253 @@
+        "//sw/device/lib/testing/... ",
+        "//sw/device/lib/runtime/...",
+        "//sw/device/lib/crypto/...",
+        "//sw/device/lib/arch/...",
+        "//sw/device/lib/ujson/...",
+        "//sw/device/lib/base:status",
+        "//sw/device/lib/base:mmio_on_device_do_not_use_directly",
+        "//sw/device/lib/base:mmio_on_host_do_not_use_directly",
+        "//sw/device/silicon_creator/rom_ext/...",
+        "//sw/device/silicon_creator/rom/keys/real:real",
+    ]
+    DEVICE_LIBS_QUERY = (f"kind(cc_library, ({' + '.join(DEVICE_LIBS_INC)}) "
+                         f"- ({' + '.join(DEVICE_LIBS_EXC)}))")
+
+    # Gather and build all device libs that can be built on host.
+    # We handle everyting in a single cquery instead of one cquery per test target since
+    # each invocation incurs an additional startup cost.
+    device_libs_all = run(BAZEL, "cquery", DEVICE_LIBS_QUERY,
+                          "--output=starlark", "--starlark:expr=target.label")
+    device_libs_incompat = run(
+        BAZEL, "cquery", DEVICE_LIBS_QUERY, "--output=starlark",
+        ("--starlark:expr=target.label "
+         "if 'IncompatiblePlatformProvider' in providers(target) "
+         "else ''"))
+    log.info(f"incompatible libraries: {pformat(device_libs_incompat)}")
+    device_libs = list(set(device_libs_all) - set(device_libs_incompat))
+    log.info(f"instrumented libraries: {pformat(device_libs)}")
+    # Force rebuild using a random action_env variable to make sure we have all the
+    # object (*.o) files that we want.
+    run(BAZEL, "build", "--config=coverage_clang",
+        f"--action_env=REBUILD={randint(0, 1000)}", *device_libs)
+
+    return device_libs
+
+
+def create_merged_library(device_libs: List[str], merged_so: Path):
+    """Create a single shared library from the given device libraries.
+
+    A coverage report can be created from multiple object files as long as they have
+    coverage data. However, unused inline functions can trigger confusing warnings and
+    functions that appear in more than one object file can cause undercounting. Linking
+    the object files of the sources that we are interested in into a single shared
+    library avoids such problems.
+
+    Args:
+        device_libs: List of device libraries whose object files will be linked into a
+        shared library.
+        merged_so: Path where to write the shared library.
+    """
+    # Gather all the object files and link a combined shared library with all the
+    # coverage data. This eliminates hash-mismatch warnings for multiple copies of
+    # unused inline functions. Note that we exclude mock_*.o since they contain
+    # duplicate symbols.
+    obj_files = run(
+        BAZEL, "cquery", f"({' + '.join(device_libs)})", "--output=starlark",
+        ("--starlark:expr='\\n'.join("
+         "[f.path for f in target.output_groups.compilation_outputs.to_list()]"
+         ")"))
+    obj_files = [o for o in obj_files if "mock" not in o]
```

Comment:
Nit: could we tighten up `"mock" not in o` as `not o.startswith("mock_")`?

Reply:
Done with two deltas:
- Renamed `mod_exp_{ibex,otbn}_mock` to `mock_mod_exp_{ibex,otbn}` everywhere
- Used `PurePath(o).name` instead of `o` since `o` is a path like `bazel-out/k8-fastbuild/bin/sw/device/silicon_creator/lib/_objs/manifest_on_device_do_not_use_directly/manifest.pic.o`